### PR TITLE
Only check strict_dependencies and layer options when --packager-layers flag is passed

### PR DESCRIPTION
**Pull request by neilparikh**
### Motivation

This provides a way to enable the packaging system, without having to enable the layering checks too.

### Test plan

See included automated tests.

### CODE REVIEW

File: core/packages/PackageDB.cc [hunk 1]
```diff
@@ -217,6 +217,10 @@ const std::vector<core::NameRef> &PackageDB::layers() const {
     return layers_;
 }
 
+const bool PackageDB::layeringChecksEnabled() const {
```
Comment (jez) [inline core/packages/PackageDB.cc:220]:
```suggestion
const bool PackageDB::enforceLayering() const {
```

to match the name of the lint rules?

completely stylistic, feel free to ignore.

File: main/options/options.cc [hunk 2]
```diff
@@ -1204,23 +1204,21 @@ void readOptions(Options &opts,
...
+                        logger->error("--packager-layers must contain items that are alphanumeric.");
+                        throw EarlyReturnWithCode(1);
+                    }
+                    opts.packagerLayers.emplace_back(layer);
                 }
-                opts.packagerLayers.emplace_back(layer);
+            } else {
+                logger->error("--packager-layers can only be specified in --stripe-packages mode");
+                throw EarlyReturnWithCode(1);
             }
```
Comment (jez) [inline main/options/options.cc:1222]: Is this nested for stylistic reasons, or is there something about the implicit_value change that meant that the `raw.count` calls don't work the same way they use to, making this structure semantically different from the old, early-return form?

File: packager/packager.cc [hunk 3]
```diff
@@ -1124,6 +1124,13 @@ struct PackageSpecBodyWalk {
         }
 
         if (send.fun == core::Names::strictDependencies()) {
+            if (!ctx.state.packageDB().layeringChecksEnabled()) {
+                if (auto e = ctx.beginError(send.loc, core::errors::Packager::InvalidStrictDependencies)) {
+                    e.setHeader("`{}` can only be declared when `{}` is passed", send.fun.show(ctx),
+                                "--packager-layers");
```
Comment (jez) [inline packager/packager.cc:1130]: Suggestion for wording:

```
Found layer annotation, but `--packager-layers` was not passed

Note: use `--packager-layers=foo,bar` to define the valid layers, or `--packager-layers` to use the default layers of `library` and `application`
```

just a suggestion

File: main/options/options.cc [hunk 4]
```diff
@@ -630,7 +630,7 @@ buildOptions(const vector<pipeline::semantic_extension::SemanticExtensionProvide
                                  "and `export` directives",
                                  cxxopts::value<vector<string>>(), "<name>");
     options.add_options(section)("packager-layers", "Valid layer names for packages, ordered lowest to highest.",
```
Comment (jez) [inline main/options/options.cc:632]: Might be worth mentioning that this both: defines the layers and opts into the checks.

**Review by jez — APPROVED** [review]

File: main/options/options.cc (hunk 2 above)
Comment (neilparikh) [reply main/options/options.cc:1222]: Previously, it was possible for `raw.count("packager-layers")` to be 0, but `raw["packager-layers"]` still have values. Now, if `raw.count(...)` is 0, then `raw[...]` will also have no values, so we need to check the count first before looping through (`raw[...]` raises an exception if the given flag doesn't have a value).

**Comment by neilparikh** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_R2nmAJSJYOZ6zu
→ https://go/builds/bui_R2nmzZ6TUETxNr
→ https://go/builds/bui_R2nmjPSXUujzmH

**Pull request merged**